### PR TITLE
Peter Gray - CS7 - WIP

### PR DESCRIPTION
**Pull request by PeterGrayCreative**
## Trello Set Up:
* [x] Use your existing Trello account from the Front End Project, or create a new one
* [x] Create a new board called "Lambda Notes(Backend) - {Your Name}"
* [x] Create lists titled `backlog`,`To Do`, `In Progress`, and `Done`
* [x] Fill in the `To Do` list with the MVP features listed below
* [x] Fill in the `backlog` list with all the extra features listed below
* [x] Share your board with the project manager that has been assigned to you.  If you have not been assigned yet, reach out to your Section Lead for guidance
* [x] Add your Trello URL to your project's README.md file.  Commit the change, push it to your repository & submit a pull request

## Backend MVP Features:
* [x] Write tests for your endpoints
* [x] Create a Database for your app
* [x] Create a Node app and connect it to your database
* [x] Create a Registration Page that allows users to create accounts for your app and sign in with email/password
* [x] Store these User Accounts in your Database
* [x] Add a Sign In Page
* [x] Store Notes in your Database
* [x] Deploy your application to the web

Build an API Endpoint in your Node App for each of the following, and connect them to your React App. These should all be read from your Database and scoped to the logged in user. Write tests for your models and your endpoints. Make sure you adequately cover use cases and errors.
* [x] Display a list of notes
* [x] Create a note with a title and content
* [x] View an existing note
* [x] Edit an existing note
* [x] Delete an existing note

### CODE REVIEW

**Comment by mixelpixel** [comment]
@PeterGrayCreative Thanks for setting up your PR with the task list and Putting the Trello board URL in the README.md 👍

File: README.md [hunk 1]
```diff
@@ -34,32 +34,9 @@ Build an API Endpoint for each of the following, and connect it to your React Ap
...
-Build an API Endpoint in your Node App for each of the following, and connect them to your React App. These should all be read from your Database and scoped to the logged in user. Write tests for your endpoints.
-* [ ] Display a list of notes
-* [ ] Create a note with a title and content
-* [ ] View an existing note
-* [ ] Edit an existing note
-* [ ] Delete an existing note
-```
+***
+# Trello Board Link
+`https://trello.com/b/1BI5xT2N`
```
Comment (mixelpixel) [inline README.md]: With markdown syntax, in particular, there is a GitHub Flavored markdown, to make this a hypertext link, you can put a description in square brackets and the link following it in parenthesis, e.g.

```md
[description of awesome website](http://awesomewebsite.com)
```

[Peter's amazing Trello board for this project](https://trello.com/b/1BI5xT2N)

You can also use `#` octothorpe's as headers, e.g. for an `<h1>` use one octothorpe, or for an `<h5>` header, use five

```md
# [Click here for Peter's amazing Trello board for this project](https://trello.com/b/1BI5xT2N)

##### [Click here for Peter's amazing Trello board for this project](https://trello.com/b/1BI5xT2N)

```

# [Click here for Peter's amazing Trello board for this project](https://trello.com/b/1BI5xT2N)

##### [Click here for Peter's amazing Trello board for this project](https://trello.com/b/1BI5xT2N)

Comment (mixelpixel) [reply README.md]: https://help.github.com/articles/basic-writing-and-formatting-syntax/

**Comment by mixelpixel** [comment]
Looks like you are making good progress and chugging along 👍 

I'm updating your Task list with `* [ ] Write tests for your endpoints`

**Comment by mixelpixel** [comment]
@PeterGrayCreative I removed the "testing" task from your Task List

File: models.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+const mongoose = require('mongoose');
+const chai = require('chai');
+const { expect } = chai;
+const sinon = require('sinon');
+
+const Note = require('./models/notes-model');
+
+describe('Notes Model', () => {
+  before((done) => {
+    mongoose.connect('mongodb://localhost/test');
```
Comment (mixelpixel) [inline models.test.js]: well... I am impressed  with your testing moxie

File: routes/api-routes.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      noteBody,
+    };
+    Note.findByIdAndUpdate(id, updatedNote)
+      .then((note) => {
+        if (note) {
+          Note.findById(note.id)
+            .then((note) => {
+              if (note) {
+                Note.find({ userId: userid })
+                  .then((notes) => {
```
Comment (mixelpixel) [inline routes/api-routes.js]: Async/Await?

File: models/user-model.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+mongooseTypes.loadTypes(mongoose, 'email');
+const Email = mongoose.SchemaTypes.Email;
+
+const SALT_ROUNDS = 11;
+
+const UserSchema = Schema(
+  {
+    username: {
+      type: Email,
```
Comment (mixelpixel) [inline models/user-model.js]: nice!

File: routes/api-routes.js (hunk 3 above)
Comment (PeterGrayCreative) [reply routes/api-routes.js]: Eh, I would question whether async/await would be any easier to read. You would still have a lot of nested if statements to handle errors.

Also, I tend to like promises more, but there is an entire layer of promises/a database call that could just be removed.

Here's with the extra onion layer removed: 😀

```server.put('/note', (req, res) => {
  const { userid } = req.headers;
  console.log(req.body);
  const { noteTitle, noteBody, id } = req.body;
  if (id && (noteTitle || noteBody)) {
    const updatedNote = {
      noteTitle,
      noteBody,
    };
    Note.findByIdAndUpdate(id, updatedNote)
      .then((note) => {
        if (note) {
          Note.find({ userId: userid })
            .then((notes) => {
              if (notes) {
                res.status(200).json(notes);
              } else res.status(404).json({ message: 'No notes found.' });
            })
            .catch((err) => res.status(400).json(err));
        } else res.status(404).json({ message: 'No note with that id.' });
      })
      .catch((err) => res.status(500).json(err));
  } else
    res.status(400).json({ error: 'Please provide an ID and Title or Body.' });
});```

**Pull request closed without merging**